src/simulator/views/map/data/flat_map.py

```diff
@@ -0,0 +1,170 @@
+
+        self._add_colour(MapData.AGENT)
+        self._add_colour(MapData.TRACE)
+        self._add_colour(MapData.GOAL)
+
+        self.square_mesh = SquareMesh(self.obstacles_data.shape[0], self.obstacles_data.shape[1])
+        self.square = self.root.attach_new_node(self.square_mesh.geom_node)
+
+        self.texture = Texture(self.name + "_texture")
+        self.texture.setup_2d_texture(self.texture_w, self.texture_h, Texture.T_unsigned_byte, Texture.F_rgba8)
+        self.square.set_texture(self.texture)
+        self.texture.set_wrap_u(Texture.WM_clamp)
+        self.texture.set_wrap_v(Texture.WM_clamp)
+        self.texture.set_clear_color(LVecBase4f(*WHITE))
+
+        self.render_obstacles()
+
+    @property
+    def dim(self) -> int:
+        return 2
+
+    def __create_lines(self, c: Colour, wfc: Colour) -> Tuple[bytes, bytes, bytes]:
+        def conv(f): return int(round(f * 255))
+        def to_pixel(c): return (conv(c.b), conv(c.g), conv(c.r), conv(c.a))
+
+        wfd2 = int(self.wf_thickness // 2)
+
+        if wfd2 != self.wf_line_cnt:
+            bc = blend_colours(wfc.with_a(wfc.a * ((self.wf_thickness / 2) - wfd2)), c)
+            tpixel = to_pixel(bc)
+        else:
+            bc = None
+        
+        pixel = to_pixel(c)
+        wfc = blend_colours(wfc, c)
+        wfpixel = to_pixel(wfc)
+
+        line = array.array('B')
+        for _ in range(wfd2):
+            line.extend(wfpixel)
+        if bc is not None:
+            line.extend(tpixel)
+        for _ in range(self.square_size - 2 * self.wf_line_cnt):
+            line.extend(pixel)
+        if bc is not None:
+            line.extend(tpixel)
+        for _ in range(wfd2):
+            line.extend(wfpixel)
+        lbytes = line.tobytes()
+
+        wfline = array.array('B')
+        for _ in range(self.square_size):
+            wfline.extend(wfpixel)
+        wflbytes = wfline.tobytes()
+
+        if bc is None:
+            tlbytes = wflbytes
+        else:
+            tline = array.array('B')
+            for _ in range(wfd2):
+                tline.extend(wfpixel)
+            tline.extend(tpixel)
+            for _ in range(self.square_size - 2 * self.wf_line_cnt):
+                tline.extend(tpixel)
+            tline.extend(tpixel)
+            for _ in range(wfd2):
+                tline.extend(wfpixel)
+            tlbytes = tline.tobytes()
+
+        self.__lines[(c, wfc)] = (lbytes, wflbytes, tlbytes)
+        return lbytes, wflbytes, tlbytes
+
+    def render_square(self, p: Point, c: Colour, wfc: Colour) -> None:
+        px, py = int(p[0]), int(p[1])
+        l, wfl, tl = self.__lines[(c, wfc)] if (c, wfc) in self.__lines else self.__create_lines(c, wfc)
+
+        img = self.texture.modify_ram_image()
+        x_offset = px * self.square_size
+        lsize = self.square_size * 4
+        for y in range(py * self.square_size, py * self.square_size + self.wf_line_cnt - 1):
```

Comment:
don't really know what these are for but you can probably clean these 3 for loops up (or add comments)